app/models/reservation.rb

```diff
@@ -2,25 +2,36 @@ class Reservation < ApplicationRecord
   belongs_to :user
   belongs_to :equipment
 
-  validate :date_available?
+  validates :reserved_date, presence: true
   validates :total, presence: true
   validates :city, presence: true
-  validates :reserved_date, presence: true
+  validate :date_available?, if: :reserved_date
 
   private
 
   def date_available?
     equipment&.dates_reserved&.each do |date|
-      date_datetime = date.to_datetime
-      reserved_datetime = reserved_date.to_datetime
-      previous_date_plus_duration = date_datetime + equipment.duration.days + 1.day
-      reserved_date_plus_duration = reserved_datetime + equipment.duration.days
-
-      if (previous_date_plus_duration > reserved_datetime && reserved_date_plus_duration > date_datetime) ||
-         date_datetime == reserved_datetime
-        errors.add(:reserved_date, 'is already taken')
-        return
+      date_time = date.to_datetime
+
+      if (outside_date(date_time) && not_overlapping(date_time)) || equal_dates(date_time)
+        errors.add(:reserved_date, 'not available')
       end
     end
   end
+
+  def outside_date(date_in_datetime)
+    date_plus_duration_plus_1_day = date_in_datetime + equipment.duration.days + 1.day
+
+    reserved_date < date_plus_duration_plus_1_day
+  end
+
+  def not_overlapping(date_in_datetime)
+    reserved_date_plus_duration = reserved_date + equipment.duration.days
+
+    reserved_date_plus_duration > date_in_datetime
+  end
+
+  def equal_dates(date_in_datetime)
+    date_in_datetime == reserved_date
+  end
 end
```

Comment:
Nice work making this more readable. One tiny thing though is the naming of the functions. Since they return a boolean, we should name them with an ```?``` at the end. Like ```outside_date?```

Reply:
Yeap, that is correct.